ree/reeEstimateError.cc

```diff
@@ -0,0 +1,458 @@
+      apf::destroyElement(el);
+      apf::destroyMeshElement(me);
+
+      // compute integral
+      double w = weight * jdet;
+      theta_plus_tk = theta_plus_tk * w;
+
+      // matrix vector multiplication
+      for (int i = 0; i < nedofs; i++)
+        lambda(i) += theta_plus_tk * tetVectorShapes[i];
+
+    } // end integral loop
+    apf::destroyMeshElement(fme);
+  } // end face loop
+}
+
+static void getEssentialElementNDDofs(apf::Mesh* mesh, apf::MeshEntity* e,
+    apf::Field* f, apf::NewArray<int>& ess_dofs, apf::NewArray<int>& uness_dofs)
+{
+  apf::FieldShape* fs = f->getShape();
+  int etype = mesh->getType(e);
+  PCU_ALWAYS_ASSERT(etype == apf::Mesh::TET);
+  int nedofs = apf::countElementNodes(fs, etype);
+
+  apf::NewArray<int> marker_list (nedofs);
+  for (int i = 0; i < nedofs; i++)
+    marker_list[i] = 0;
+
+  // populate marker list by iterating over downward edges and faces
+  apf::Downward edges;
+  int nedges = mesh->getDownward(e, 1, edges);
+  PCU_ALWAYS_ASSERT(nedges == 6);
+  for (int i = 0; i < nedges; i++) {
+    int nodesOnEdge = fs->countNodesOn(mesh->getType(edges[i]));
+    if ( isOnDomainBoundary(mesh, edges[i]) ) {
+      for (int n = 0; n < nodesOnEdge; n++) {
+        marker_list[(i*nodesOnEdge)+n] = -1;
+      }
+    }
+  }
+  int nodesOnEdges = fs->countNodesOn(apf::Mesh::EDGE) * nedges;
+
+  apf::Downward faces;
+  int nfaces = mesh->getDownward(e, 2, faces);
+  PCU_ALWAYS_ASSERT(nfaces == 4);
+  for (int i = 0; i < nfaces; i++) {
+    int nodesOnFace = fs->countNodesOn(mesh->getType(faces[i]));
+    if ( isOnDomainBoundary(mesh, faces[i]) ) {
+      for (int n = 0; n < nodesOnFace; n++) {
+        marker_list[(nodesOnEdges + i*nodesOnFace)+n] = -1;
+      }
+    }
+  }
+
+  int num_marked = 0;
+  for (int i = 0; i < nedofs; i++) {
+    if (marker_list[i])
+      num_marked++;
+  }
+
+  // use marker list to get ess_dofs list
+  ess_dofs.allocated() ? ess_dofs.resize(num_marked)
+    : ess_dofs.allocate(num_marked);
+  uness_dofs.allocated() ? uness_dofs.resize(nedofs-num_marked)
+    : uness_dofs.allocate(nedofs-num_marked);
+  int ess_dof_counter = 0;
+  int uness_dof_counter = 0;
+  for (int i = 0; i < nedofs; i++) {
+    if(marker_list[i])
+      ess_dofs[ess_dof_counter++] = i;
+    else
+      uness_dofs[uness_dof_counter++] = i;
+  }
+}
+
+/**
+ * Inputs: Matrix A, Vector X, Vector B, essential dofs, other dofs.
+ * Output: reduced matrix A, reduced rhs B.
+ */
+static void eliminateDBCs(
+    mth::Matrix<double> const &A,
+    mth::Vector<double> const &X,
+    mth::Vector<double> const &B,
+    apf::NewArray<int>  const &ess_dofs,
+    apf::NewArray<int>  const &uness_dofs,
+    mth::Matrix<double> &Anew,
+    mth::Vector<double> &Bnew)
+{
+  int num_ess_dofs = ess_dofs.size();
+  int num_uness_dofs = uness_dofs.size();
+
+  // 1. Remove rows of A corresponding to
+	// ess_dofs by copying into Anew
+  Anew.resize(num_uness_dofs, num_uness_dofs);
+	for(int rr = 0; rr < num_uness_dofs; rr++) {
```

Comment:
Please replace tabs with spaces.  There are several in this file.

Reply:
Done.

Reply:
There are still tabs in this file.  Please replace them all.

vim: https://alvinalexander.com/blog/post/linux-unix/vi-vim-faq-how-replace-tab-character-space-spaces/
emacs: https://www.systutorials.com/replacing-tabs-with-spaces-in-emacs/

Reply:
Is there a way to highlight tabs to see where they exist in a file? For now, I used the `vim` link you gave to replace tab with spaces, but I am not certain if all of them have been taken care of.

Reply:
The vim search command `/\t` will find tabs in the file.  They are all gone with latest commit.